### PR TITLE
Improve challenge visibility control

### PR DESCRIPTION
**Pull request by TheOsch**

### CODE REVIEW

File: actions/challenges.js [hunk 1]
```diff
@@ -862,31 +865,47 @@ function validateChallenge(api, connection, dbConnectionMap, challengeId, isStud
...
-                    api.dataAccess.executeQuery('check_user_challenge_accessibility', sqlParams, dbConnectionMap, cbx);
-                },
-                exists:  function (cbx) {
-                    api.dataAccess.executeQuery('check_challenge_exists', sqlParams, dbConnectionMap, cbx);
-                }
-            }, cb);
+            api.dataAccess.executeQuery('get_challenge_accessibility_and_groups', sqlParams, dbConnectionMap, cb);
         }, function (res, cb) {
-            if (res.exists.length === 0 || Boolean(res.exists[0].is_studio) !== isStudio) {
+            // If the record with this callengeId doesn't exist in contest_eligibility table
```
Comment (skyhit) [inline actions/challenges.js]: @TheOsch it is possible that there is no records, which mean the challenge is pubic group.

**Comment by skyhit** [comment]
@TheOsch for https://github.com/appirio-tech/tc-api/blob/master/actions/challengeRegistration.js#L81, can we do the similar check as validation?

File: actions/challenges.js (hunk 1 above)
Comment (TheOsch) [inline actions/challenges.js]: Fixed.

**Comment by TheOsch** [comment]
We discussed it with Tony Jefts yesterday. In fact, the validateChallenge function (where the link in the challenge spec points to) is used only in actions that get checkpoints. 
So there are many more fragments in the code besides the registration that need to be changed too. 
We agreed that i'll make the PR with only verifyChallenge updated just to check the approach itself, but it will not be the end of the job, and next we will find all places in the code that need to be changed and fix them. Including registration.

**Comment by skyhit** [comment]
@TheOsch I see

**Comment by TheOsch** [comment]
@skyhit I looked at the list of actions at routes.js and wrote done all that look like in scope. When you'll have time would you please briefly look at this list and prompt me if there are some actions in the list that have nothing in common with our task, or maybe I missed some? I think you know the API much better than I, and it will save us a lot of time.

 get: [
        { path: "/:apiVersion/challenges/registrants/:challengeId", action: "getRegistrants" },
        { path: "/:apiVersion/challenges/submissions/:challengeId/mySubmissions", action: "getUserSubmissions" },
        { path: "/:apiVersion/challenges/submissions/:challengeId", action: "getSubmissions" },
        { path: "/:apiVersion/challenges/phases/:challengeId", action: "getPhases" },
        { path: "/:apiVersion/challenges/active", action: "getActiveChallenges" },
        { path: "/:apiVersion/challenges/open", action: "getOpenChallenges" },
        { path: "/:apiVersion/challenges/upcoming", action: "getUpcomingChallenges" },
        { path: "/:apiVersion/challenges/past", action: "getPastChallenges" },
        { path: "/:apiVersion/challenges/:challengeId", action: "getChallenge" },
        { path: "/:apiVersion/challenges", action: "searchSoftwareAndStudioChallenges" },
        { path: "/:apiVersion/develop/challenges/result/:challengeId", action: "getSoftwareChallengeResults" },
        { path: "/:apiVersion/develop/challenges/:challengeId", action: "getSoftwareChallenge" },
        { path: "/:apiVersion/develop/challenges", action: "searchSoftwareChallenges" },
        { path: "/:apiVersion/develop/reviewOpportunities/:challengeId", action: "getSoftwareReviewOpportunity" },
        { path: "/:apiVersion/develop/reviewOpportunities", action: "searchReviewOpportunities" },
        { path: "/:apiVersion/design/challenges/result/:challengeId", action: "getStudioChallengeResults" },
        { path: "/:apiVersion/design/reviewOpportunities/:id", action: "getStudioReviewOpportunity" },
        { path: "/:apiVersion/design/challenges/:challengeId", action: "getStudioChallenge" },
        { path: "/:apiVersion/design/challenges", action: "searchStudioChallenges" },
        { path: "/:apiVersion/design/reviewOpportunities", action: "getStudioReviewOpportunities" },
        { path: "/:apiVersion/design/download/:submissionId", action: "downloadDesignSubmission" },
        { path: "/:apiVersion/software/reviewers/:contestType", action: "getChallengeReviewers" },
    ],
    post: [
        { path: "/:apiVersion/develop/reviewOpportunities/:challengeId/apply", action: "applyDevelopReviewOpportunity" },
        { path: "/:apiVersion/develop/challenges/:challengeId/submit", action: "submitForDevelopChallenge" },
        { path: "/:apiVersion/develop/challenges/:challengeId/upload", action: "uploadForDevelopChallenge" },
        { path: "/:apiVersion/design/challenges/:challengeId/submit", action: "submitForDesignChallenge" },
        { path: "/:apiVersion/design/reviewOpportunities/:challengeId/apply", action: "applyDesignReviewOpportunity" },
    ]

**Comment by skyhit** [comment]
@ajefts since we are using v3 apis, I think some of the apis above are not used anymore.

**Comment by ajefts** [comment]
@TheOsch Where does TC_API_V3_URL get read from?

**Comment by ajefts** [comment]
@skyhit Yes, let's review the list when you have time.  We won't need to update all of those.

**Comment by TheOsch** [comment]
@ajefts 

> Where does TC_API_V3_URL get read from?

From the environment. For example, one can run the command like
`set TC_API_V3_URL=https://real.api.topcoder.com/v3/` (in Windows) or `export TC_API_V3_URL=https://real.api.topcoder.com/v3/` (in Unix) before `npm start`.

>  let's review the list

Meanwhile i'll work on challengeRegistration.js . Already began working on it, will commit tomorrow.

**Comment by TheOsch** [comment]
Committed. Sorry for a delay. What's next?

**Comment by skyhit** [comment]
@TheOsch we will test the current logic in dev, then we can decide the next step

**Comment by skyhit** [comment]
@TheOsch can you future make the changes to `getSoftwareChallenge` related logic?

**Comment by skyhit** [comment]
@TheOsch mainly logic related to the following check

https://github.com/appirio-tech/tc-api/blob/e8d29a7ec36e6d1336c31505b01de2dcb3255205/actions/challenges.js#L3354-L3355

**Comment by TheOsch** [comment]
@skyhit Easily. Changing both actions (getSoftwareChallenge and getRegistrants) will take one day, I can make it by tomorrow.

**Comment by TheOsch** [comment]
@skyhit Merged changes you made into my own fork, updated and made another PR.
https://github.com/appirio-tech/tc-api/pull/504

**Pull request merged**